### PR TITLE
Remove NuGet references from packages library

### PR DESCRIPTION
**Pull request by twsouthwick**
This moves the final usages of NuGet.* assemblies out of the packages class. This adds a couple of abstractions to help with this as well.

### CODE REVIEW

File: src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetVersionComparer.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.UpgradeAssistant.MSBuild
+{
+    public class NuGetVersionComparer : IVersionComparer
+    {
+        public int Compare(string? x, string? y)
+        {
+            if (NuGetVersion.TryParse(x, out var nx))
```
Comment (mjrousos) [inline src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetVersionComparer.cs:12]: It looks like we might be missing a `!` here (and on the next case)

**Review by mjrousos — CHANGES_REQUESTED** [review]
Looks good; a few changes requested around checking Compare direction and things like that.

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/PackageMapReferenceAnalyzer.cs [hunk 2]
```diff
@@ -70,6 +74,34 @@ public async Task<PackageAnalysisState> AnalyzeAsync(IProject project, PackageAn
             return state;
         }
 
+        /// <summary>
+        /// Determines whether a package map's .NET Framework packages include a
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/PackageMapReferenceAnalyzer.cs:78]: I don't see anything that would necessarily be framework specific

File: src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetVersionComparer.cs (hunk 1 above)
Comment (twsouthwick) [reply src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetVersionComparer.cs:12]: Added tests

File: src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetExtensions.cs [hunk 3]
```diff
@@ -1,20 +1,53 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics.CodeAnalysis;
 using NuGet.Versioning;
 
 namespace Microsoft.DotNet.UpgradeAssistant.MSBuild
 {
-    internal static class NuGetExtensions
+    public static class NuGetExtensions
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/NuGetExtensions.cs:9]: It's for tests that I moved from the packages lib

File: src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/PackageMapReferenceAnalyzer.cs [hunk 4]
```diff
@@ -70,6 +74,34 @@ public async Task<PackageAnalysisState> AnalyzeAsync(IProject project, PackageAn
...
+            }
+
+            // If the version isn't specified, then matching the name is sufficient
+            // Similarly, if the NetFx package has a wildcard version, then any version matches
+            if (version is null || reference.HasWildcardVersion)
+            {
+                return true;
+            }
+
+            return _comparer.Compare(version, reference.Version) > 0;
```
[the remark this answers is not in the document]
Comment (twsouthwick) [reply src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/Analyzers/PackageMapReferenceAnalyzer.cs:102]: For some reason I keep getting the result of compare backwards. I just reviewed again on MSDN .... hopefully it sticks

**Review by mjrousos — APPROVED** [review]
Looks good

**Pull request merged**